src/middlewared/middlewared/plugins/truenas_connect/acme.py

```diff
@@ -0,0 +1,197 @@
+        private = True
+        namespace = 'tn_connect.acme'
+
+    async def call(self, url, mode, payload=None):
+        config = await self.middleware.call('tn_connect.config_internal')
+        return await self._call(url, mode, payload=payload, headers=await self.auth_headers(config))
+
+    async def config(self):
+        config = await self.middleware.call('tn_connect.config_internal')
+        creds = get_account_id_and_system_id(config)
+        if not config['enabled'] or creds is None:
+            return {
+                'error': 'TrueNAS Connect is not enabled or not configured properly',
+                'tnc_configured': False,
+                'acme_details': {},
+            }
+
+        resp = await self.call(ACME_CONFIG_URL.format(account_id=creds['account_id']), 'get')
+        resp['acme_details'] = resp.pop('response')
+        if resp['error'] is None:
+            resp = normalize_acme_config(resp)
+
+        return resp | {
+            'tnc_configured': True,
+        }
+
+    async def update_ui(self):
+        logger.debug('Updating UI with TNC cert')
+        try:
+            await self.update_ui_impl()
+        except Exception:
+            logger.error('Failed to configure TNC cert in the UI', exc_info=True)
+            await self.middleware.call('tn_connect.set_status', Status.CERT_CONFIGURATION_FAILURE.name)
+        else:
+            logger.debug('TNC cert configured successfully')
+            await self.middleware.call('tn_connect.set_status', Status.CONFIGURED.name)
+            # Let's restart UI now
+            # TODO: Hash this out with everyone
+            await self.middleware.call('system.general.ui_restart', 2)
+
+    async def update_ui_impl(self):
+        config = await self.middleware.call('tn_connect.config')
+        # We are going to set the cert now for system UI
+        if config['certificate'] is None:
+            # Just some sanity testing
+            raise CallError('Certificate is not set for TrueNAS Connect')
+        await self.middleware.call('system.general.update', {'ui_certificate': config['certificate']})
+
+    async def initiate_cert_generation(self):
+        logger.debug('Initiating cert generation steps for TNC')
+        try:
+            cert_details = await self.initiate_cert_generation_impl()
+        except Exception:
+            logger.error('Failed to complete certificate generation for TNC', exc_info=True)
+            await self.middleware.call('tn_connect.set_status', Status.CERT_GENERATION_FAILED.name)
+        else:
+            cert_id = await self.middleware.call(
+                'datastore.insert',
+                'system.certificate', {
+                    'name': f'truenas_connect_{str(uuid.uuid4())[-5:]}',
+                    'type': CERT_TYPE_EXISTING,
+                    'certificate': cert_details['cert'],
+                    'privatekey': cert_details['private_key'],
+                    'renew_days': CERT_RENEW_DAYS,
```

Comment:
Does this constant need to be defined? Renewal times should come from the CA

Reply:
They are coming from the CA itself, this is how many days before the cert itself expires that we renew it. So if it expires on 31st and this value is 2, so we would attempt to renew it 2 days before 31st comes

Reply:
I see, so 15 days is just a default?